src/storage/src/source/persist_source.rs

```diff
@@ -132,91 +128,40 @@ where
             let mut current_ts = 0;
 
             move |cap, output| {
-                let waker = futures_util::task::waker_ref(&activator);
                 let mut context = Context::from_waker(&waker);
-
-                // Bound execution of operator to prevent a single operator from hogging
-                // the CPU if there are many messages to process
-                let timer = Instant::now();
-
                 while let Poll::Ready(item) = pinned_stream.as_mut().poll_next(&mut context) {
                     match item {
-                        Some(Ok(event)) => match event {
-                            ListenEvent::Progress(upper) => {
-                                // NOTE(aljoscha): Ideally, we would only get a CapabilitySet, that
-                                // we have to manage. Or the ability to drop the plain Capability
-                                // right at the start.
-                                if let Some(first_element) = upper.first() {
-                                    current_ts = *first_element;
-                                    cap.downgrade(first_element);
-                                }
-                            }
-                            ListenEvent::Updates(updates) => {
-                                let cap = cap.delayed(&current_ts);
-                                let mut session = output.session(&cap);
-                                for update in updates {
-                                    session.give(Ok(update));
-                                }
+                        Some(Ok(ListenEvent::Progress(upper))) => match upper.into_option() {
+                            Some(ts) => {
+                                current_ts = ts;
+                                cap.downgrade(&ts);
                             }
+                            None => return SourceStatus::Done,
                         },
-                        Some(Err::<_, anyhow::Error>(e)) => {
+                        Some(Ok(ListenEvent::Updates(mut updates))) => {
                             let cap = cap.delayed(&current_ts);
                             let mut session = output.session(&cap);
-                            session.give(Err((format!("{}", e), current_ts, 1)));
+                            session.give_vec(&mut updates);
                         }
-                        None => {
-                            unreachable!("We poll from persist continuously, the Stream should therefore never be exhausted.")
+                        Some(Err::<_, ExternalError>(e)) => {
+                            // TODO(petrosagg): error handling
+                            panic!("unexpected error from persist {e}");
                         }
-                    }
-
-                    if timer.elapsed() > YIELD_INTERVAL {
```

Comment:
Hm, I might have answered this too soon. In this case we're using the tokio reactor but the timely executor and it's unclear who does the accounting. I'll try to look into this

Reply:
Yeah, this looks like a tokio scheduler optimization which does not come into play here, I will re-add this

Reply:
I was wondering, because you were the one who added it in the raw source operator :sweat_smile: 

https://github.com/MaterializeInc/materialize/blob/1fe9848c28ef6ebdfbce673c629a877a5287143e/src/storage/src/source/mod.rs#L1091